From 28ec7168e490444d5355715c11d2cada9885c5dc Mon Sep 17 00:00:00 2001
From: Tom White <tom@cloudera.com>
Date: Wed, 19 Sep 2012 10:56:27 +0100
Subject: [PATCH 1197/1357] MR1: MAPREDUCE-4470. Fix TestCombineFileInputFormat.testForEmptyFile

Reason: Regression
Ref: CDH-8046
Author: Ilya Katsov
(cherry picked from commit 7ee3100b5814f304b34f936aa442ca6af47020d8)
---
 .../hadoop/mapred/lib/CombineFileInputFormat.java  |    3 ++
 .../mapred/lib/TestCombineFileInputFormat.java     |   25 ++++++++++++++++++++
 2 files changed, 28 insertions(+), 0 deletions(-)

diff --git a/src/mapred/org/apache/hadoop/mapred/lib/CombineFileInputFormat.java b/src/mapred/org/apache/hadoop/mapred/lib/CombineFileInputFormat.java
index 7b6f8c6..7d3c722 100644
--- a/src/mapred/org/apache/hadoop/mapred/lib/CombineFileInputFormat.java
+++ b/src/mapred/org/apache/hadoop/mapred/lib/CombineFileInputFormat.java
@@ -466,6 +466,9 @@ public abstract class CombineFileInputFormat<K, V>
       if (locations == null) {
         blocks = new OneBlockInfo[0];
       } else {
+        if(locations.length == 0) {
+          locations = new BlockLocation[] { new BlockLocation() };
+        }
         blocks = new OneBlockInfo[locations.length];
         for (int i = 0; i < locations.length; i++) {
            
diff --git a/src/test/org/apache/hadoop/mapred/lib/TestCombineFileInputFormat.java b/src/test/org/apache/hadoop/mapred/lib/TestCombineFileInputFormat.java
index ce5fce2..c02a44e 100644
--- a/src/test/org/apache/hadoop/mapred/lib/TestCombineFileInputFormat.java
+++ b/src/test/org/apache/hadoop/mapred/lib/TestCombineFileInputFormat.java
@@ -33,6 +33,7 @@ import org.apache.hadoop.hdfs.DFSTestUtil;
 import org.apache.hadoop.hdfs.DistributedFileSystem;
 import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.fs.PathFilter;
+import org.apache.hadoop.mapred.FileInputFormat;
 import org.apache.hadoop.mapred.InputSplit;
 import org.apache.hadoop.mapred.JobConf;
 import org.apache.hadoop.mapred.Reporter;
@@ -452,6 +453,30 @@ public class TestCombineFileInputFormat extends TestCase{
     stm.close();
     DFSTestUtil.waitReplication(fileSys, name, replication);
   }
+  
+  public void testForEmptyFile() throws Exception {
+    Configuration conf = new Configuration();
+    FileSystem fileSys = FileSystem.get(conf);
+    Path file = new Path("test" + "/file");
+    FSDataOutputStream out = fileSys.create(file, true,
+        conf.getInt("io.file.buffer.size", 4096), (short) 1, (long) BLOCKSIZE);
+    out.write(new byte[0]);
+    out.close();
+
+    // split it using a CombinedFile input format
+    DummyInputFormat inFormat = new DummyInputFormat();
+    JobConf job = new JobConf(conf);
+    FileInputFormat.setInputPaths(job, "test");
+    InputSplit[] splits = inFormat.getSplits(job, /*ignored*/ 7);
+    assertEquals(1, splits.length);
+    CombineFileSplit fileSplit = (CombineFileSplit) splits[0];
+    assertEquals(1, fileSplit.getNumPaths());
+    assertEquals(file.getName(), fileSplit.getPath(0).getName());
+    assertEquals(0, fileSplit.getOffset(0));
+    assertEquals(0, fileSplit.getLength(0));
+
+    fileSys.delete(file.getParent(), true);
+  }
 
   static class TestFilter implements PathFilter {
     private Path p;
-- 
1.7.0.4

